sol/private/sol_binary.bzl

```diff
@@ -0,0 +1,175 @@
+    solinfo = ctx.toolchains["@aspect_rules_sol//sol:toolchain_type"].solinfo
+    args = ctx.actions.args()
+
+    # User-provided arguments first, so we can override them
+    args.add_all(ctx.attr.args)
+
+    args.add_all([s.path for s in ctx.files.srcs])
+
+    # TODO: is this the right value? maybe it ought to be the package directory?
+    args.add_all(["--base-path", "."])
+
+    args.add("--output-dir")
+    args.add_joined([ctx.bin_dir.path, ctx.label.package], join_with = "/")
+
+    root_packages = []
+    for dep in ctx.attr.deps:
+        if JsInfo in dep:
+            for pkg in dep[JsInfo].transitive_npm_linked_packages.to_list():
+                # Where the node_modules were installed
+                root_packages.append(pkg.store_info.root_package)
+        if SolSourcesInfo in dep:
+            for prefix, target in dep[SolSourcesInfo].remappings.items():
+                args.add_joined([prefix, target], join_with = "=")
+
+    if len(root_packages):
+        args.add("--include-path")
+        args.add_joined(
+            root_packages,
+            format_each = ctx.bin_dir.path + "/%s/node_modules",
+            join_with = ",",
+            uniquify = True,
+        )
+
+    if ctx.attr.bin:
+        args.add("--bin")
+    if ctx.attr.ast_compact_json:
+        args.add("--ast-compact-json")
+    for v in ctx.attr.combined_json:
+        if v not in _OUTPUT_COMPONENTS:
+            fail("Illegal output component {}, must be one of {}".format(v, _OUTPUT_COMPONENTS))
+    if len(ctx.attr.combined_json):
+        args.add("--combined-json")
+        args.add_joined(ctx.attr.combined_json, join_with = ",")
+
+    outputs = _calculate_outs(ctx)
+    if not len(outputs):
+        fail("No outputs were requested. This is illegal under Bazel, as actions are only run to produce output files.")
+
+    npm_deps = js_lib_helpers.gather_files_from_js_providers(ctx.attr.deps, include_transitive_sources = True, include_declarations = False, include_npm_linked_packages = True)
+
+    # solc will follow symlinks out of the sandbox, then insist that the execroot path is allowed.
+    #
+    # This seems to be an understood limitation, for example in https://github.com/ethereum/solidity/issues/11410:
+    # > NOTE: --allowed-directories becomes almost redundant after these changes. There are now only two cases where it's needed:
+    # > When a file is a symlink that leads to a file outside of base path or include directories.
+    # > The directory containing the symlink target must be in --allowed-directories for this to be allowed.
+    #
+    # Effectively disable this security feature - Bazel's sandbox ensures reproducibility
+    # Anyhow, as very few compilers do such a thing, the solc layer isn't the right place to solve.
+    args.add_all(["--allow-paths", "/"])
+
+    shim = ctx.actions.declare_file("_solc_{}.sh".format(ctx.label.name))
+    ctx.actions.write(shim, """#!/usr/bin/env bash
```

Comment:
nit. put this in bash script into its own file and use write_source_files trick to write out an expanded one to sources to it can be shell checked. we did in rules_js.

Reply:
actually this whole shim can be removed, it turns out we don't need to get solc output via stdout at this point, and so the only remaining function is to make it silent on success, and that can go in the new wrap_binary in bazel-lib